src/Core/Cart/CartRow.php

```diff
@@ -245,7 +245,7 @@ public function getFinalTotalPrice()
     public function processCalculation(Cart $cart)
     {
         $rowData = $this->getRowData();
-        $quantity = (int) $rowData['cart_quantity'];
+        $quantity = (int) $rowData['quantity'];
```

Comment:
This change is puzzling, because here you're using the stock quantity instead of the number of items in the cart (see https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Cart.php#L642).

Reply:
@eternoendless

It must be overwritten later in the code.

Quantity = quantity of items of that row.
Cart_quantity = total quantity of that product in cart.

Make a cart rule with gift and apply it. The price of the product is 100.
Then, add 3 more pieces of that product normally.

Without gift separation:
- cart_quantity = 4
- quantity = 4
Everything is fine. Total is 4x100 = 400

With gift separation enabled:
Row 1 - cart_quantity 4, quantity 1
Row 2 - cart_quantity 4, quantity 3
Total is 4x100 + 4x100 = 800 = wrong

After changing cart_quantity to quantity, its 1x100 + 3x100 = 400 = correct

Reply:
> It must be overwritten later in the code.

Found it! https://github.com/PrestaShop/PrestaShop/blob/1.7.6.0/classes/Cart.php#L842-L844